packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+          setState(() {
+            _rootBucket = bucket;
+            _rootBucketValid = true;
+            _okToRenderBlankContainer = false;
+          });
+        }
+        RendererBinding.instance.allowFirstFrame();
+      });
+    }
+  }
+
+  void _replaceRootBucket() {
+    _rootBucketValid = false;
+    _rootBucket = null;
+    ServicesBinding.instance.restorationManager.removeListener(_replaceRootBucket);
+    _loadRootBucketIfNecessary();
+    assert(!_isWaitingForRootBucket); // Ensure that load finished synchronously.
+  }
+
+  @override
+  void dispose() {
+    ServicesBinding.instance.restorationManager.removeListener(_replaceRootBucket);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    if (_okToRenderBlankContainer && _isWaitingForRootBucket) {
+      return Container();
+    }
+
+    return UnmanagedRestorationScope(
+      bucket: _ancestorBucket ?? _rootBucket,
+      child: RestorationScope(
+        restorationId: widget.restorationId,
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+/// Manages an object of type `T`, whose value a [State] object wants to have
+/// restored during state restoration.
+///
+/// The property wraps an object of type `T`. It knows how to store its value in
+/// the restoration data and it knows how to re-instantiate that object from the
+/// information it previously stored in the restoration data.
+///
+/// The knowledge of how to store the wrapped object in the restoration data is
+/// encoded in the [toPrimitives] method and the knowledge of how to
+/// re-instantiate the object from that data is encoded in the [fromPrimitives]
+/// method. A call to [toPrimitives] must return a representation of the wrapped
+/// object that can be serialized with the [StandardMessageCodec]. If any
+/// collections (e.g. [List]s, [Map]s, etc.) are returned, they must not be
+/// modified after they have been returned from [toPrimitives]. At a later point
+/// in time (which may be after the application restarted), the data obtained
+/// from [toPrimitives] may be handed back to the property's [fromPrimitives]
+/// method to restore it to the previous state described by that data.
+///
+/// A [RestorableProperty] needs to be registered to a [RestorationMixin] using
+/// a [RestorationId] that is unique within the mixin. The [RestorationMixin]
+/// provides and manages the [RestorationBucket], in which the data returned by
+/// [toPrimitives] is stored.
+///
+/// Whenever the value returned by [toPrimitives] (or the [enabled] getter)
+/// changes, the [RestorableProperty] must call [notifyListeners]. This will
+/// trigger the [RestorationMixin] to update the data it has stored for the
+/// property in its [RestorationBucket] to the latest information returned by
+/// [toPrimitives].
+///
+/// When the property is registered with the [RestorationMixin], one of two
+/// things can happen: The mixin checks whether there is any restoration data
```

Comment:
I was reading this paragraph, looking for the "other thing", thinking that the first thing was the "The mixin checks", but after reading it again, I think that the "one of two things" part should come after the "The mixin checks" sentence, so that the first thing is the case of finding data, and the second thing is not finding data.

Reply:
I rephrased this.